### ISSUE TITLE
Fix S2589: FP with conditional access in previous condition

### ISSUE THREAD

**Issue by Thieum** [SonarSource/sonar-dotnet#2164]
### Description

rule S2589 will give a false positive with conditional access in previous condition in cascading ifs.

### Repro steps

```
using System;
using static System.Console;

public class Result
{
    public bool Succeed { get; set; }

    public static Result Test()
    {
        if (DateTime.Now.Day == 17) // swap value here to test both cases if needed
        {
            return new Result();
        }
        return null;
    }
}
internal static class Program
{
    internal static void Main()
    {
        var result = Result.Test();

        if (result == null || !result.Succeed)
        {
            WriteLine("shorted");
            if (result != null)
            {
                WriteLine("other");
            }
        }

        if (result?.Succeed != true)
        {
            WriteLine("shorted");
            if (result != null)
            {
                WriteLine("other");
            }
        }

        ReadKey();
    }
}
```

### Expected behavior

The two series of tests should be equivalent, and sonar should treat them as such.

### Actual behavior

Sonar will pop a S2589 on the second ifs in the second serie of ifs.

### Known workarounds

do not use conditional access in cascading ifs, or silence S2589.

### Related information

* SonarLint for Visual Studio 4.7.0.3891
* Visual Studio 2017 15.9.4

**Comment by Thieum** [comment]
Still an issue with SonarLint for Visual Studio 4.8.0.4040

**Comment by Evangelink** [comment]
Hi @Thieum,

As you can see your ticket is still open and not assigned to any milestone which means it wasn't yet handled.

Thank you for your patience!

Cheers,
Amaury

**Comment by Thieum** [comment]
@Evangelink No worries, and thanks for the heads-up! Just seen the new version had a lot of fixes, so could have been fixed through a border effect. Thanks for a great tool :)

**Comment by andrei-epure-sonarsource** [comment]
I confirm it's a FP on SonarC# 7.13

**Comment by andrei-epure-sonarsource** [comment]
Probably #2361 , #2164 and #1837 are all related to how we interpret the null-conditional operator

**Comment by andrei-epure-sonarsource** [comment]
Adding also an example from #2407

```
class A
{
  public bool booleanVal { get; set; }
}

A aObj = null;
if (aObj?.booleanVal ?? false)
```

**Comment by andrei-epure-sonarsource** [comment]
I removed this from 7.15 milestone as there should be a separate milestone dedicated to fixing advanced rules which use Symbolic Execution

**Comment by Thieum** [comment]
Another example:

```
if (_aiController?.IsConnected == true)
{
   Logger.Instance.Log("AI Controller is already connected.");
   return;
 }

 if (_aiController != null)  // <= S2589 FP
 {
    Logger.Instance.Log("Disposing old connection ai...");
    _aiController.Dispose();
    _aiController = null;
}
```

**Comment by andrei-epure-sonarsource** [comment]
The last test added by @Thieum is related to #1347 (the way we consider that `IsNullOrEmpty` is the same with `IsNull`)

**Comment by andrei-epure-sonarsource** [comment]
fixed in #2497
